components/board/genericlinux/gpio.go

```diff
@@ -79,24 +99,89 @@ func (pin *gpioPin) Get(ctx context.Context, extra map[string]interface{}) (resu
 	return (value != 0), nil
 }
 
+// Lock the mutex before calling this! We'll spin up a background goroutine to create a PWM signal
+// in software, if one isn't already running.
+func (pin *gpioPin) startSoftwarePWM() {
+	if pin.pwmRunning {
+		// We're already running a software PWM loop, we don't need another.
+		return
+	}
+	pin.pwmRunning = true
+	pin.waitGroup.Add(1)
+	goutils.ManagedGo(pin.softwarePwmLoop, pin.waitGroup.Done)
+}
+
+// We turn the pin either on or off, and then wait until it's time to turn it off or on again (or
+// until we're supposed to shut down). We return whether we should continue the software PWM cycle.
+func (pin *gpioPin) halfPwmCycle(shouldBeOn bool) bool {
```

Comment:
Tell me more. What were you hoping to optimize? I could inline this function into the two places that call it, at a cost of doubling the number of places where bugs need to be fixed (but maybe the extra maintenance costs are worth it in here). I could move the calculations of how long the duration is to outside of the lock, but that only frees up the lock for other threads, without making this any faster. There's probably some way to get rid of the if statement on lines 125-126 to get rid of the memory fence there, but I don't have a sense of whether it would be faster or not to have something like this instead:
```go
shouldBeOnInt := int(shouldBeOn)
dutyCycle := shouldBeOnInt * pin.pwmDutyCyclePct + (1 - shouldBeOnInt) * (1 - pin.pwmDutyCyclePct)
```